Fix last seen unconfirmed

ValuedMammal opened the pull request:
A method `update_last_seen_unconfirmed` is added for `TxGraph` that allows inserting a `seen_at` time for all unanchored transactions.

fixes #1336 

### Changelog notice

Changed:
- Methods `into_tx_graph` and `into_confirmation_time_tx_graph`for `RelevantTxids` are changed to no longer accept a `seen_at` parameter.

Added:
- Added method `update_last_seen_unconfirmed` for `TxGraph`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

notmandatory commented:
This looks fine to me, but my understanding is that a time value always should be provided, even if it's made up as long the the values are increasing. If we make this optional we should at least explain what the downside is of using None, which I believe was unneeded DB updates.

ValuedMammal commented:
> This looks fine to me, but my understanding is that a time value always should be provided, even if it's made up as long the the values are increasing. If we make this optional we should at least explain what the downside is of using None, which I believe was unneeded DB updates.

Ok, I think I agree it should be required.

rustaceanrob commented:
I agree on required. It should also make for a better UX as the end user "refreshes" their wallet.

evanlinjin commented:
Sorry about the conflicts! We have merged #1171

ValuedMammal commented:
electrum: `into_tx_graph` for RelevantTxids takes an optional `seen_at`, that could be why I was undecided before whether it should always be provided.

@evanlinjin Can we assume in electrum_ext if a relevant txid has no associated anchors, it must be unconfirmed and therefore be an appropriate time to `insert_seen_at`?

evanlinjin commented:
@ValuedMammal that sounds right to me! Let's add a couple of tests and I'll be happy to ACK and merge.

**My proposed tests for both `ElectrumExt` and `EsploraExt`:** When syncing, ensure that transactions that stay unconfirmed between rounds have their `last_seen` increased and transactions that stay confirmed have unchanging `last_seen`s.

ValuedMammal commented:
Update on test coverage:

For electrum, repeated syncing doesn't update the last seen timestamp, because if a tx is already present in graph, TxGraph won't report it as missing.

For esplora I did successfully check that timestamps increase for repeat syncs as long as a tx is unconfirmed. When the tx confirms, the timestamp is unchanged.

Note: I don't love the fact that every test needs to mine 100 regtest blocks, as that slows down CI. We should perhaps think about consolidating various assertions into more monolithic tests.

LLFourn commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -190,3 +193,57 @@ fn tx_can_become_unconfirmed_after_reorg() -> Result<()> {
+    let _ = env.mine_empty_block();
+    env.wait_until_electrum_sees_block()?;
+    let ElectrumUpdate { relevant_txids, .. } = client.sync(chain.tip(), [spk], None, None, 5)?;
+    let missing = relevant_txids.missing_full_txs(graph.graph());
```
we want to avoid writing these kinds of tests that use `apply_update` and `missing_full_txs` because we're going to change the mechanisms on how this works. Tests against the blockchain backends should just check that it returns the right data (not that it is right after applying it as an update).

ValuedMammal replied:
You're right, it's a bit overkill.

LLFourn reviewed (commented):
The changes here look exactly how I imagined they should. Thanks! But now that I see it, isn't it simpler just to have method like `update_last_seen_unconfirmed` on `TxGraph` which sets the last seen of all unanchored transactions? We can just call this on the update `TxGraph` and set it to `now` before applying it. This means we wouldn't need to pass it in, we can just mutate the `TxGraph` afterwards.

At least presently it doesn't look like we lose anything by doing this since the backends are both just setting whatever you tell them when you pass the time in. I was thinking that maybe esplora or electrum would tell you the last time it was seen in the API but I guess that's not the case. Mempool.space does have this api: https://mempool.space/docs/api/rest#get-transaction-times but that's "first seen" rather than last seen.

ValuedMammal commented:
> The changes here look exactly how I imagined they should. Thanks! But now that I see it, isn't it simpler just to have method like `update_last_seen_unconfirmed` on `TxGraph` which sets the last seen of all unanchored transactions? We can just call this on the update `TxGraph` and set it to `now` before applying it. This means we wouldn't need to pass it in, we can just mutate the `TxGraph` afterwards.

I'll have another look at this.

ValuedMammal committed:
feat(tx_graph): Add method update_last_seen_unconfirmed

That accepts a `u64` as param representing the latest timestamp
and internally calls `insert_seen_at` for all transactions in
graph that aren't yet anchored in a confirmed block.

ValuedMammal commented:
- Updated the PR description to reflect the approach using `TxGraph`, so the above comments are mostly outdated. This way has the advantage of not breaking the current API for electrum/esplora.

notmandatory commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -46,8 +46,10 @@ impl RelevantTxids {
-            if let Some(seen_at) = seen_at {
-                let _ = graph.insert_seen_at(txid, seen_at);
+            if anchors.is_empty() {
+                if let Some(seen_at) = seen_at {
```
❓ I know it breaks the current API but shouldn't `seen_at` be required and non-optional here too?

ValuedMammal replied:
I don't know whether it makes more sense to make this `seen_at` parameter required, or to remove the param and use the new `update_last_seen_unconfirmed` method on the update `TxGraph`.

LLFourn replied:
I think it's fine as it is. Will likely be radically changed soon.

[EDIT] but when you did:
```rust
 l    let mut graph_update =
        relevant_txids.into_confirmation_time_tx_graph(&client, None, missing)?;
    let now = std::time::UNIX_EPOCH.elapsed().unwrap().as_secs();
    let _ = graph_update.update_last_seen_unconfirmed(now);
```
You are choosing to pass in `None` but then manually update it. If this is what you want to do in the examples then just remove the option from these functions. It's confusing to use both in the examples.

ValuedMammal replied:
Fixed. I made `seen_at` a required parameter for `into_tx_graph`.

LLFourn reviewed (changes requested):
LGTM just need to make the electrum examples consistent (decide whether to pass it into the `into_confirmation_time_tx_graph` or just remove the Option and do it manually outside).

evanlinjin reviewed (commented):
This looks great! However, there may be a situation where the chain-source returns anchors that do not point to a block in the best chain. Those unconfirmed transactions will therefore have no "timestamp". Would it be better if `TxGraph::update_last_seen_unconfirmed` takes in a checkpoint tip so unconfirmed transactions are transactions that do not have an anchor that points to a checkpoint?

ValuedMammal committed:
fix(electrum)!: Remove `seen_at` param from `into_tx_graph`

and `into_confirmation_time_tx_graph`, since now it makes
sense to use `TxGraph::update_last_seen_unconfirmed`.

Also, use `update_last_seen_unconfirmed` in examples for
electrum/esplora. We show how to update the last seen
time for transactions by calling `update_last_seen_unconfirmed`
on the graph update returned from a blockchain source, passing
in the current time, before applying it to another `TxGraph`.

ValuedMammal commented:
> This looks great! However, there may be a situation where the chain-source returns anchors that do not point to a block in the best chain. Those unconfirmed transactions will therefore have no "timestamp". Would it be better if `TxGraph::update_last_seen_unconfirmed` takes in a checkpoint tip so unconfirmed transactions are transactions that do not have an anchor that points to a checkpoint?

I had not thought of that

LLFourn commented:
> This looks great! However, there may be a situation where the chain-source returns anchors that do not point to a block in the best chain. Those unconfirmed transactions will therefore have no "timestamp". Would it be better if `TxGraph::update_last_seen_unconfirmed` takes in a checkpoint tip so unconfirmed transactions are transactions that do not have an anchor that points to a checkpoint?

`last_seen_unconfirmed` is for transactions that we've actually seen as unconfirmed in a fully validating mempool. Why would we set it for transactions that we've only seen as confirmed? Or are you saying esplora will allow us to determine that a tx is both in the mempool but was also confirmed in some non-canonical block? If that's the case then we have to pass in the timestamp into the sync/scan function so the logic can set both. The suggestion that we add and use `TxGraph::update_last_seen_unconfirmed` was based on the assumption that esplora and electrum would not allow us to distinguish this case.

evanlinjin commented:
@LLFourn the problem is we can't get blocks and transactions atomically. We either get blocks before getting transactions, or get transactions before getting blocks.

There might be a situation when a reorg happens between fetching txs and blocks. The tx might be sent back to mempool because it's not included in replacement block.

LLFourn commented:
> @LLFourn the problem is we can't get blocks and transactions atomically. We either get blocks before getting transactions, or get transactions before getting blocks.
> 
> There might be a situation when a reorg happens between fetching txs and blocks. The tx might be sent back to mempool because it's not included in replacement block.

...or it might not be if the re-org confirms a double spend. It might be in the mempool it might not be so we don't set `last_seen` because we don't know. Where's the problem?

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -549,6 +549,32 @@ impl<A: Clone + Ord> TxGraph<A> {
+    /// Update the last seen unconfirmed time for all unanchored transactions.
+    ///
+    /// Returns the [`ChangeSet`] after applying any updates to `self`. Note that
+    /// [`TxGraph`] only keeps track of the latest `seen_at`.
```
It'd be nice if the docs explain more about why you would want to use this method. Are there some docs about how "last seen" works around that we could link to?

One question that comes to mind is why isn't it good enough to just set the `last_seen` to the most recent block header time. An example, showing how to get the right time from `std::time` would be good.

ValuedMammal replied:
Pushed a docs only commit, let me know what you think.

@evanlinjin Is there a preference for keeping `into_tx_graph` as shown in c551dfba304464cf0814a9211b66d34e513f53b7 versus dropping the `seen_at` parameter in favor of the new `update_last_seen_unconfirmed`?

Also, passing in a chain tip is an interesting idea that could be its own method if we could demonstrate a use for it, but isn't `TxGraph` supposed to be agnostic to whether a tx is "canonical"?

LLFourn replied:
docs are :100:. Your docs explain nicely why we don't want to use the block header time.

evanlinjin replied:
@ValuedMammal do you mean dropping the `seen_at` parameter from `into_tx_graph`? I think that will be nice thanks!

LLFourn reviewed (approved):
ACK b21ea0fe53710cf5ae037d9da48d48355e9f654e

From looking at it, the changes in `electrum_ext.rs` seem to do exactly the same thing that would be achieved if you just called `update_last_seen_unconfirmed` so it would make sense to just do that instead of passing in `seen_at`. I'm pretty sure @evanlinjin will change this soon so I don't think it matters that much.

LLFourn reviewed (approved):
ACK 449dd9adec6aa95babfb339ab7ce815e5f7e0e74

evanlinjin reviewed (commented):
LGTM! Thanks for the quality work. Let's drop the `last_seen` parameter from `.into_tx_graph` and I'll ACK/merge.

ValuedMammal commented:
> LGTM! Thanks for the quality work. Let's drop the `last_seen` parameter from `.into_tx_graph` and I'll ACK/merge.

Done. Thank you to the reviewers @evanlinjin @LLFourn @notmandatory @rustaceanrob

evanlinjin reviewed (approved):
ACK a2a64ffb6e92baf46a912f36294f3f4f521a528a
